app/badges.py

```diff
@@ -1,29 +1,120 @@
+    We also want to be able to create badges.
 
-    "shitposter2018": {"nick": "2018shit", "name": "Shitposter of the year", "alt": _l("Winner of the shitposter of the year 2018 contest"), "icon": "shitposter18.svg", "score": 250}
+    For backwards compatability we will allow "fetching" of old_badges but only by ID.
 
-}
+    This will also create an interfact for Triggers, as Badges and Triggers are interlinked.
+    """
+
+    def __iter__(self):
+        """
+        Returns a list of all bagdes in the database.
+        """
+        return (x for x in Badge.select(Badge.bid, Badge.name, Badge.alt, Badge.icon, Badge.score, Badge.trigger, Badge.rank)\
+                .order_by(Badge.rank, Badge.name))
+
+    def __getitem__(self, bid):
+        """
+        Returns a badge from the database.
+        """
+        try:
+            return Badge.get(Badge.bid == bid)
+        except:
+            return None
+
+    def update_badge(self, bid, name, alt, icon, score, rank, trigger):
+        """
+        Updates the information related to a badge, updates icon if provided.
+        """
+        if icon:
+            icon = gen_icon(icon)
+        else:
+            icon = self[bid].icon
+
+        Badge.update(name=name, alt=alt, icon=icon, score=score, rank=rank, trigger=trigger).where(Badge.bid==bid).execute()
+
+
+    def new_badge(self, name, alt, icon, score, rank, trigger=None):
+        """
+        Creates a new badge with an optional trigger.
+        """
+        icon = gen_icon(icon)
+        Badge.create(name=name, alt=alt, icon=icon,
+                     score=score, rank=rank, trigger=trigger)
+
+    def delete_badge(self, bid):
+        """
+        Deletes a badge by ID
+        """
+        Badge.delete().where(Badge.bid == bid).execute()
+        UserMetadata.delete().where((UserMetadata.key == 'badge')
+                                    & (UserMetadata.value == bid)).execute()
+
+    def assign_userbadge(self, uid, bid):
+        """
+        Gives a badge to a user
+        """
+        UserMetadata.get_or_create(key="badge", uid=uid, value=bid)
+
+    def triggers(self):
+        """
+        Lists available triggers that can be attached to a badge.
+        """
+        return triggers.keys()
 
-for bg in badges:
-    try:
-        badges[bg]['icon'] = open('./app/static/svg/' + badges[bg]['icon']).read()
-    except FileNotFoundError:
-        pass
+    def badges_for_user(self, uid):
+        """
+        Returns a list of badges associated with a user.
+        """
+        return Badge.select(Badge.bid, Badge.name, Badge.icon, Badge.score, Badge.alt, Badge.rank)\
+            .join(UserMetadata, JOIN.LEFT_OUTER, on=(UserMetadata.value == Badge.bid.cast("varchar")))\
```

Comment:
What's the reason for the `.cast("varchar")`? This seems to break in MySQL/MariaDB.

Reply:
So, Badge.bid is a int, and so you can't just compare an int to a string in postgres, so you have to cast it.